current_proto.go

```diff
@@ -0,0 +1,117 @@
+package main
+
+import (
+	"os"
+
+	"github.com/emicklei/proto"
+)
+
+type ProtoMessageMap map[string]*ProtoMessage
+
+func BuildCurrentProtoMap(filename string) (ProtoMessageMap, error) {
+	if filename == "" {
+		return ProtoMessageMap{}, nil
+	}
+	reader, _ := os.Open(filename)
+	defer reader.Close()
+
+	parser := proto.NewParser(reader)
+	definition, _ := parser.Parse()
+
+	p := make(ProtoMessageMap)
+	proto.Walk(definition, proto.WithMessage(p.HandleMessage()))
+
+	return p, nil
+}
+
+func (p ProtoMessageMap) HandleMessage() func(m *proto.Message) {
+	return func(m *proto.Message) {
+		p[m.Name] = NewProtoMesssageFromMessage(m)
+	}
+}
+
+func (p ProtoMessageMap) GetFieldNum(messageName, fieldName string) int {
+	_, ok := p[messageName]
+	if !ok {
+		p[messageName] = &ProtoMessage{}
+	}
+	fieldNum := p[messageName].GetFieldNum(fieldName)
+	return fieldNum
+}
+
+func (p ProtoMessageMap) RemoveFieldNum(messageName, fieldName string) {
+	_, ok := p[messageName]
+	if !ok {
+		return
+	}
+	p[messageName].RemoveFieldNum(fieldName)
+	return
+}
+
+type ProtoMessage struct {
+	currMaxNum  int
+	droppedNums []int
+	fields      map[string]int
+}
+
+func NewProtoMesssageFromMessage(msg *proto.Message) *ProtoMessage {
+	fieldsMap := make(map[string]int, len(msg.Elements))
+	currMax := 0
+
+	for i, _ := range msg.Elements {
+		element := msg.Elements[i]
+
+		switch element.(type) {
+		case *proto.NormalField:
+			field := element.(*proto.NormalField)
+			fieldsMap[field.Name] = field.Sequence
+			if field.Sequence > currMax {
+				currMax = field.Sequence
+			}
+		case *proto.MapField:
+			field := element.(*proto.MapField)
+			fieldsMap[field.Name] = field.Sequence
+			if field.Sequence > currMax {
+				currMax = field.Sequence
+			}
+		default:
+		}
+	}
+	return &ProtoMessage{
+		currMaxNum: currMax,
+		fields:     fieldsMap,
+	}
+}
+
+func (p *ProtoMessage) GetFieldNum(fieldName string) int {
+	if p.fields == nil {
+		p.fields = make(map[string]int)
+	}
+	_, ok := p.fields[fieldName]
+	if ok {
+		return p.fields[fieldName]
+	}
+
+	if len(p.droppedNums) == 0 {
+		p.currMaxNum++
+		p.fields[fieldName] = p.currMaxNum
+		return p.fields[fieldName]
+	}
+
+	p.fields[fieldName] = p.droppedNums[0]
```

Comment:
Does this assume order? If I call fieldNames out of order that dropped num was created would my map be wrong? If that is the case maybe add a comment or use a map for droppedNums instead of a slice?

Reply:
It does assume order. I'll add a comment